File: backend/src/shared/property-monetization.ts
// shared/property-monetization.ts
//
// Free-forever residential types vs subscription-limited types.

/** Catalog keys that are 100% free forever (unlimited create, ads-supported). */
export const FREE_RESIDENTIAL_TYPES = [
  'rental', // Rental Property
  'house',  // Independent House
  'villa',  // Villa
] as const;

/** Apartment is subscription-limited: 1 building + 20 rooms/flats on Free. */
export const APARTMENT_TYPE = 'apartment';

export const FREE_APARTMENT_MAX_BUILDINGS = 1;
export const FREE_APARTMENT_MAX_ROOMS = 20;

/** Level internal_keys that count as "buildings" for apartment free limits. */
export const BUILDING_LEVEL_KEYS = ['building', 'tower'] as const;

/** Level internal_keys that count as "rooms" for apartment free limits. */
export const ROOM_LEVEL_KEYS = ['room', 'flat', 'bed'] as const;

export type FreeResidentialType = (typeof FREE_RESIDENTIAL_TYPES)[number];

export function isFreeResidentialType(propertyTypeKey: string): boolean {
  return (FREE_RESIDENTIAL_TYPES as readonly string[]).includes(propertyTypeKey);
}

export function isApartmentType(propertyTypeKey: string): boolean {
  return propertyTypeKey === APARTMENT_TYPE;
}

/** Types that get 1 free property, then require paid for more. */
export function isSubscriptionLimitedPropertyType(propertyTypeKey: string): boolean {
  return !isFreeResidentialType(propertyTypeKey);
}

export function isBuildingLevelKey(internalKey: string): boolean {
  return (BUILDING_LEVEL_KEYS as readonly string[]).includes(internalKey);
}

export function isRoomLevelKey(internalKey: string): boolean {
  return (ROOM_LEVEL_KEYS as readonly string[]).includes(internalKey);
}

/** Paid / unlimited access — not on the free plan. */
export function isPaidSubscription(
  planSlug: string | null | undefined,
  status: string | null | undefined,
): boolean {
  if (!planSlug || planSlug === 'free') return false;
  return status === 'active' || status === 'trialing' || status === 'past_due';
}

export function monetizationMetaForType(propertyTypeKey: string) {
  if (isFreeResidentialType(propertyTypeKey)) {
    return { tier: 'free_forever', badge: 'Free with Ads', freeLimit: null };
  }
  if (isApartmentType(propertyTypeKey)) {
    return {
      tier: 'apartment',
      badge: '1 Building + 20 Rooms free',
      freeLimit: { buildings: FREE_APARTMENT_MAX_BUILDINGS, rooms: FREE_APARTMENT_MAX_ROOMS },
    };
  }
  return { tier: 'commercial', badge: '1 Free, Upgrade for Multi-property', freeLimit: 1 };
}

export class SubscriptionRequiredError extends Error {
  readonly statusCode = 403;
  readonly code = 'SUBSCRIPTION_REQUIRED';

  constructor(message = 'Upgrade required to create another commercial property.') {
    super(message);
    this.name = 'SubscriptionRequiredError';
  }
}

File: mobile/lib/features/properties/domain/property_monetization.dart
// features/properties/domain/property_monetization.dart
//
// Mirrors backend/src/shared/property-monetization.ts.

/// Free forever (unlimited create, ads-supported).
const kFreeResidentialTypes = {
  'rental', // Rental Property
  'house', // Independent House
  'villa', // Villa
};

const kApartmentType = 'apartment';

/// Free Apartment structure caps (paid plans unlimited).
const kFreeApartmentMaxBuildings = 1;
const kFreeApartmentMaxRooms = 20;

const kBuildingLevelKeys = {'building', 'tower'};
const kRoomLevelKeys = {'room', 'flat', 'bed'};

class PropertyMonetization {
  static bool isFreeResidential(String? propertyTypeKey) {
    if (propertyTypeKey == null || propertyTypeKey.isEmpty) return false;
    return kFreeResidentialTypes.contains(propertyTypeKey);
  }

  static bool isApartment(String? propertyTypeKey) =>
      propertyTypeKey == kApartmentType;

  /// Apartment + Hostel/PG/etc. — not free-forever residential.
  static bool isCommercial(String? propertyTypeKey) {
    if (propertyTypeKey == null || propertyTypeKey.isEmpty) return true;
    return !isFreeResidential(propertyTypeKey);
  }

  static bool isPaidPlan(String? planSlug, String? status) {
    if (planSlug == null || planSlug == 'free') return false;
    return status == 'active' || status == 'trialing' || status == 'past_due';
  }

  static bool isBuildingLevel(String? internalKey) =>
      internalKey != null && kBuildingLevelKeys.contains(internalKey);

  static bool isRoomLevel(String? internalKey, {bool supportsOccupancy = false}) =>
      supportsOccupancy ||
      (internalKey != null && kRoomLevelKeys.contains(internalKey));

  static int get freeApartmentMaxBuildings => kFreeApartmentMaxBuildings;
  static int get freeApartmentMaxRooms => kFreeApartmentMaxRooms;

  static String badgeFor(String? propertyTypeKey) {
    if (isFreeResidential(propertyTypeKey)) return 'Free with Ads';
    if (isApartment(propertyTypeKey)) return '1 Building + 20 Rooms free';
    return '1 Free, Upgrade for Multi-property';
  }

  /// Whether creating another property of [typeKey] should open the paywall.
  static bool requiresUpgradeForCreate({
    required String? typeKey,
    required int commercialCount,
    required bool isPaid,
  }) {
    if (isFreeResidential(typeKey)) return false;
    if (isPaid) return false;
    return commercialCount >= 1;
  }

  /// Message shown when tapping a frozen surplus commercial property.
  static const expiredLockMessage =
      'Your Premium plan has expired. Upgrade to access all your properties.';

  /// Client-side lock: free residential always open; oldest commercial open;
  /// other commercial locked when not paid. Prefer server `is_locked` when present.
  static bool isPropertyLocked({
    required Map<String, dynamic> property,
    required List<Map<String, dynamic>> allProperties,
    required bool isPaid,
  }) {
    if (isPaid) return false;

    final serverLocked = property['is_locked'];
    if (serverLocked is bool) return serverLocked;
    final serverManageable = property['manageable'];
    if (serverManageable is bool) return !serverManageable;

    final typeKey = property['property_type_key'] as String?;
    if (isFreeResidential(typeKey)) return false;

    final commercials = allProperties
        .where((p) => isCommercial(p['property_type_key'] as String?))
        .toList()
      ..sort((a, b) {
        final aAt = DateTime.tryParse('${a['created_at'] ?? ''}') ??
            DateTime.fromMillisecondsSinceEpoch(0);
        final bAt = DateTime.tryParse('${b['created_at'] ?? ''}') ??
            DateTime.fromMillisecondsSinceEpoch(0);
        return aAt.compareTo(bAt);
      });

    if (commercials.isEmpty) return false;
    return property['id'] != commercials.first['id'];
  }
}
